service/defaultconfig/config.go

```diff
@@ -0,0 +1,44 @@
+// Copyright 2020 The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package defaultconfig
+
+import (
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configmodels"
+)
+
+// CreateDefaultConfig creates default configuration for a given configuration.
+func CreateDefaultConfig(factories component.Factories, forConfig *configmodels.Config) *configmodels.Config {
+	// the default config cannot be installed to the missing pipeline
+	// because the service will fail to start as there is no default exporter.
+	if forConfig == nil || forConfig.Service.Pipelines["traces"] == nil {
```

Comment:
I had to add `forConfig` param for the reason described ^^ e.g. the service would fail to start if there was defined only metrics pipeline. 

I am not sure if I like this API now. Maybe we should not return the default config but instead modify the supplied one by installing missing components. Now is the challenge what we should do with the receivers, processors array from the pipeline. Shall we not touch it if the user defined it (this allows disabling default components)?

Reply:
Yeah, this is unfortunate. I agree with you. Also the behavior is hard to explain to the users and can be a source of confusion. We may actually make it worse by trying to make the user's life easier.

The more I think about it the less appealing I find the initially idea of the degault config. I am inclined to give up on the notion of the default config unless we can come up with a clear and intuitive behavior. I have a hard time suggesting a good approach myself.

Reply:
I was more talking here from the internal API perspective. I still see benefits in the default config. Users shouldn't be required to always configure batching, memory limiter, the minimal user interaction the better.

For the end-user the rationale should be:
* the  default config should not cause start failure, any failures
* the default config should be overridable
* user should know the final configuration

Reply:
Sure, I agree that the concept of the default config is generally useful. But I don't like how our particular implementation of it works from the user's perspective.

The workaround you have on line 27 is not going to work if the traces pipeline has a different name (which we allow). Even if it worked, how do we rationalize that default config only applies when there is a traces pipeline but not when there is a metrics pipeline only?

Reply:
> The workaround you have on line 27 is not going to work if the traces pipeline has a different name (which we allow)

We could make it work by checking the pipeline type and properly injecting the configuration. But I agree that it might be cryptic.